### ISSUE TITLE
PB Sources Coding Conventions

### ISSUE THREAD

**Issue by tajmone** [fantaisie-software/purebasic#5]
A few consideration on the status of PB sources in this repository, and potential problems with PureBasic IDE settings that might modify the repository contents.

# Missing BOM

While sifting through the sources of this project I noticed that PB sources don't have a BOM. This might become a problem in view of users' contributions, especially if they code with PureBasic IDE for it enforces the BOM on UTF-8 files (and using the Ascii setting as a fallback would corrupt the encoding).

The main problem here is that just by saving a file in the PB IDE it would introduce the BOM, which would be seen by Git as an unstaged change. I stongly advice adding the BOM to all PB sources, for it won't harm Ascii files and will prevent future problems with Unicode — after all, using the BOM is the official way of handling PB sources in Unicode.

- [x] Add a BOM to all sources.
- [x] Set `.editorconfig` to handle all PB sources as `charset = utf-8-bom`.
- [x] Check that files are correctly encoded (i.e. didn't get corrupted during Git operations due to lack of BOM)
    + [x] If any files are encoding-corrupted, try to fix them with dedicate tools (iconv, etc.)

Right now, [EditorConfig] settings for PB files must be set to:

    charset =  utf-8

whereas the proper  setting should be:

    charset = utf-8-bom

If we fix the problem by adding a BOM to all source, we can then enforce `utf-8-bom` on all PB files via `.editorconfig`.

# Code Linting

Linting is another potential source of problems, because I noticed that in a few places the code uses custom indentation which would be lost by reformatting the code using the PB IDE indentation formatter.

Agreeing on a standard linting for PB code is a good way to prevent bad diffs and spurious commits. I think that using PB IDE's native formatter as _the_ reference linter would be a good solution since every user owns it and knows it. So if it's OK with you Fred, it might be worth sacrificing custom indentation in favour of adopting the PB IDE indentation formatter for the sake of standardizing PB linting rules.

- [x] Reformat all PB sources using PureBasic IDE (menu: __Edit__ » __Format indentation__, <kbd>Ctrl</kbd>+<kbd>I</kbd>)

The only problem with this is that the PB IDE adds indentation to empty lines between indented code. For example (spaces represented with `•`):

```purebasic
If 5 > 0
••
  Debug "True!"
••
EndIf
```

This prevents exploiting the [EditorConfig] setting `trim_trailing_whitespace = true` because it would see those indented empty lines as an error — in my EditorConfig PR I've set it to `unset` to prevent [EClint] from checking for trailing whitespace. Although extra whitespace is not a huge issue, it would be nice to be able to cover it in coding style conventions.

## Creating a Standalone Linter from PB IDE Code

A solution that came to mind is to isolate the PB IDE sources that deal with indentation reformatting and create a standalone tool which could then be used to lint source files in this repository. We could then ask contributors to manually run this linter before submitting changes, or even set a Git filter to run the linter at commit time (we'd only need to keep the compiled binaries inside the repo, so Git can find them on all OSs).

## PB IDE: Allow Disabling Indentation of Empty Lines

It might also be a good idea to improve the IDE by adding an option in the IDE Settings that prevents indentation of empty lines, which would simplify enforcing code consistency via [EditorConfig]. Of course, there's nothing wrong with indenting empty lines according to the surrounding code, many editors do it and it actually makes editing new insertions easier; but sometimes compromises are needed for the sake of better integration with standards and tools.

# PB IDE Settings Guidelines

There are many pitfalls when working with Git on PureBasic projects using the PB IDE with the "wrong" settings — i.e. all those settings which store personal data or configurations in the source files, which lead to spurious changes in the repository.

- [ ] Add `CONTRIBUING.md` guidelines on how to properly set the PB IDE for working on/with the repository without disrupting Git workflow.
- [ ] Create some tutorials on the project's Wiki on how to set up the PB IDE correctly (with more details than the `CONTRIBUING.md` guidelines).

I've written a few articles and posts on this issue around GitHub:

- [PB Project and Source Files]  — Issue #11 at MLF4PB/MLF-Dev.
- [Source Settings Save Options] — Wiki page at tajmone/PBasmUI.
- [Git Resources for PureBASIC] — at [tajmone/pb-archives-test], with assets.

I'm planning to create a series of new articles and tutorials on Git and PB project in the [_PureBasic Repository Template_ Wiki]. That material might be a useful reference for contributors to this project, Git configurations are rather complex, especially to Git newbies, and correct settings require experience of the various pitfalls of collaborative editing across different OSs, starting from EOL normalization and many other small details tha could easily turn the collaborative experience into a nightmare for users new to Git and GitHub.

I'd be happy to share my material to create some contributors guideline for this project, either as markdown documents inside the repo or by using the project Wiki.

You can count on me for anything concerning documentation, which is my field of expertise (I've worked in digital publishing in the past, and took on software documentation as a personal passion).

[EditorConfig]: https://editorconfig.org/
[EClint]: https://www.npmjs.com/package/eclint
[Travis CI]: https://travis-ci.com/

[tajmone/pb-archives-test]: https://github.com/tajmone/pb-archives-test/tree/source/pb-development/git
[Git Resources for PureBASIC]: https://htmlpreview.github.io/?https://github.com/tajmone/pb-archives-test/blob/source/pb-development/git/index.html "Live HTML Preview"
[PB Project and Source Files]: https://github.com/MLF4PB/MLF-Dev/issues/11
[Source Settings Save Options]: https://github.com/tajmone/PBasmUI/wiki/Source-Settings-Save-Options

[_PureBasic Repository Template_ Wiki]: https://github.com/tajmone/purebasic-repository-template/wiki

**Comment by tajmone** [comment]
## EditorConfig and Empty Lines Indentation

Regarding the above mentioned problem of EditorConfig and PureBasic IDE indenting empty lines, I've discovered that there's already a new feature request discussion on EditorConfig to allow this type of indentation without considering it trailing whitespace:

- editorconfig/editorconfig#238

Hopefully, this feature request will be approved and implemented, which would allow us to benefit from `trim_trailing_whitespace` validation on PB sources compliant to how the PB IDE formats code.

**Comment by tajmone** [comment]
> May be the best would be to add a BOM to every .pb/pbi file and use bom-utf8 in EditorConfig ?

Definitely, I agree with this. Also, I propose that every PB file (`.pb`, `.pbi`, `.pbf`) is opened with the latest PB IDE and reformatted (via <kbd>Crtl</kbd>+<kbd>I</kbd>) so that all files adopt a standard linting convention (which reduces spurious hunks in commits and improves diffs readability).

But, as I mentioned in #5, this would flatten out some arbitrary indentation sporadically found in the sources (mainly, after labels and nested compiler directives); but if it's OK with you to sacrifice custom styles in favour of standard linting we could ensure contributors an easy way to format code — all PB users have the PB IDE installed with the language. I edit PB files mainly in Sublime Text, but I always use the PB IDE for linting, for it does a great job in getting indentation and alignments right, as well as fixing keywords casing.

> We wrote the IDE and tool with the IDE didn't support UTF8 so it remained like that, but now it should use BOM or we can't use the PB IDE to edit files. Could you amend your commit to add BOM ?

I'll fix the issue and either amend the commit (if you don't mind me force pushing) or add a new one.

If you agree on reformatting every file with a BOM, then I'll open every single PB source and reformat it using the IDE. It's a bit time consuming but it's well worth the effort, especially since some users are already starting to work on IDE improvements and unless we set a standard right away we'll end up having to face contributions with messy diffs which are hard to exmaine due to excessive hunks with whitespace differences.

**Comment by fantaisie-software** [comment]
OK makes sense

**Comment by tajmone** [comment]
Done! I've validated all `*.pb` files using `isutf8` (under MSYS2):

    $ find . -name \*.pb | xargs isutf8 -l

No files are reported broken (also tried with `-i` instead of `-l`, and all files are reported as good).

So everything seems fine. I remember that I came across some weird characters being misrepresented when the BOM was not there (even if my editor should default to UTF8), but can't remember exactly were. So, my guess is that everything is fine with the BOM back in place.
